Add rename primary and foreign key operation

Brar opened the pull request:
Addresses #5662 

~~I've kept separate commits for the three different constraint types but will squash them upon request.~~

- [X] The code builds
- [X] All but 2 tests pass (verified by running `.\build.cmd -test` on my system). The 2 failing tests seem to be completely unrelated to my changes as they also fail on a clean master checkout on my system:
  - `Microsoft.EntityFrameworkCore.Scaffolding.SqlServerDatabaseModelFactoryTest.Default_value_matching_clr_default_is_not_stored` fails for `IgnoredDefault24 float NOT NULL DEFAULT 0.0E0` where `null` is expected and the actual value is `0.000000000000000e+000`
  - `Microsoft.EntityFrameworkCore.Scaffolding.SqlServerDatabaseModelFactoryTest.Hidden_columns_are_not_created` throws an `SqlException` : `Incorrect syntax near 'GENERATED'.\r\nIncorrect syntax near 'HIDDEN'.\r\nIncorrect syntax near the keyword 'with'. If this statement is a common table expression, an xmlnamespaces clause or a change tracking context clause, the previous statement must be terminated with a semicolon.`
- [ ] Commit messages currently don't follow this format as they aren't finalized yet
```
    Summary of the changes
    - Detail 1
    - Detail 2

    Fixes #bugnumber
```
- [X] Tests for the changes have been added (for bug fixes / features)
- [X] Code meets the expectations our engineering guidelines. https://github.com/aspnet/Home/wiki/Engineering-guidelines.

Please review the guidelines for CONTRIBUTING.md for more details.

Brar commented:
Sorry this was meant to be a *Draft* pull request but I accidentally created a normal one and it appears that you cant change this after it has been created (I've actually already [bugged](https://twitter.com/BrarPiening/status/1096467579806330881) @natfriedman about this on twitter).

Brar commented:
I've rebased this because it didn't apply cleanly anymore.
As rebasing while keeping several commits in order is a pain, I've also squashed the commits into one single commit.

bricelam commented on src/EFCore.Relational/Migrations/MigrationsSqlGenerator.cs:
```diff
@@ -372,6 +375,27 @@ protected virtual IUpdateSqlGenerator SqlGenerator
+            [CanBeNull] IModel model,
+            [NotNull] MigrationCommandListBuilder builder)
+        {
+            throw new NotImplementedException();
```
This base implementation should decompose this back into a drop and create operation and generate the SQL for those. Providers can then override this to use something like sp_rename if it's supported.

bricelam replied:
For reference, the SQLite provider already does this for RenameIndexOperation:

https://github.com/aspnet/EntityFrameworkCore/blob/99fb2d4e3d52f8a3be67fe6815a4fc288e63fe50/src/EFCore.Sqlite.Core/Migrations/SqliteMigrationsSqlGenerator.cs#L230-L264

Brar replied:
I agree.
This makes sense and is pretty safe and easy to implement.
I'll change it accordingly.

bricelam commented on src/EFCore.Relational/Migrations/MigrationsSqlGenerator.cs:
```diff
@@ -393,6 +417,48 @@ protected virtual IUpdateSqlGenerator SqlGenerator
+            [CanBeNull] IModel model,
+            [NotNull] MigrationCommandListBuilder builder)
+        {
+            throw new NotImplementedException();
```
Ditto

Brar replied:
I'm not so sure in this case (and in the case of `RenameUniqueConstraintOperation`).
While it could be implemented the same way as `RenameForeignKeyOperation` and `RenameIndexOperation` in contrast to them it is not guaranteed to succeed.
If there is a foreign key pointing to the primary key or unique constraint the operation will fail.

Next to the current solution of simply not supporting this if renaming isn't supported by the db provider there are two possible alternatives:

1. Just try to to it and fail if there are foreign keys pointing here. This would effectively mean "implement it the same way as `RenameForeignKeyOperation`".
2. Try to avoid the failure by trying to look up all foreign keys that point to this primary key/unique constraint and drop/recreate all of them. This would go well beyond what I'd consider "pretty safe and easy to implement".

bricelam commented on src/EFCore.Relational/Migrations/MigrationsSqlGenerator.cs:
```diff
@@ -393,6 +417,48 @@ protected virtual IUpdateSqlGenerator SqlGenerator
+            [CanBeNull] IModel model,
+            [NotNull] MigrationCommandListBuilder builder)
+        {
+            throw new NotImplementedException();
```
Ditto

Brar replied:
Same as above (`RenamePrimaryKeyOperation`).

bricelam commented on src/EFCore.SqlServer/Migrations/SqlServerMigrationsSqlGenerator.cs:
```diff
@@ -345,6 +345,34 @@ protected override void Generate(MigrationOperation operation, IModel model, Mig
+
+            if (string.IsNullOrEmpty(operation.Schema))
+            {
+                throw new InvalidOperationException(SqlServerStrings.ForeignKeySchemaRequired);
```
Won't this be `null` if it's in the default schema? We don't want to generate C# that's just going to throw...

We should handle this case. We can get the user's default schema in Transact-SQL and use EXEC to call sp_rename using a variable.

Brar replied:
Yeah, this looks like a good idea.
I'll go for it.

Brar replied:
> We can get the user's default schema in Transact-SQL and use EXEC to call sp_rename using a variable.

While drafting an implementation for this I just learned that the [documentation for `sp_rename`](https://docs.microsoft.com/en-us/sql/relational-databases/system-stored-procedures/sp-rename-transact-sql) is misleading by stating "If the object to be renamed is a constraint, object_name must be in the form schema.constraint".
At least on my local instance (SQL Server 2016 - Windows) omitting the schema part works perfectly.
I assume that this is the case in all SQL Server versions that Entity Framework Core is supposed to support, but I probably can't confirm this.
If it was, it would greatly simplify the implementation and also make it more consistent with other internal APIs where the schema part is optional.
Any chance I can get confirmation for this from inside MS?

bricelam commented on src/EFCore.SqlServer/Migrations/SqlServerMigrationsSqlGenerator.cs:
```diff
@@ -345,6 +345,34 @@ protected override void Generate(MigrationOperation operation, IModel model, Mig
+                operation.NewName,
+                "OBJECT",
+                builder);
+            builder.EndCommand(suppressTransaction: IsMemoryOptimized(operation, model, operation.Schema, operation.Table));
```
Is the `IsMemoryOptimized` check really needed? (It probably is; I just don't want to copy-and-paste it blindly)

Brar replied:
This is indeed blind copy-paste. I'll try to find out what would be the correct solution here.

bricelam replied:
Confirmed, we need this. Funny, indexes aren't supported on memory-optimized tables. Guess we'll be ready if they ever are. 😉

bricelam commented on src/EFCore.Sqlite.Core/Migrations/SqliteMigrationsSqlGenerator.cs:
```diff
@@ -497,6 +497,39 @@ protected override void Generate(AlterColumnOperation operation, IModel model, M
+        /// <param name="model"> The target model which may be <c>null</c> if the operations exist without a model. </param>
+        /// <param name="builder"> The command builder to use to build the commands. </param>
+        protected override void Generate(RenameForeignKeyOperation operation, IModel model, MigrationCommandListBuilder builder)
+            => throw new NotSupportedException(
```
These should just fallback to the base Relational implementation (which will just call the Add/Drop overrides)

Brar replied:
Yes, for foreign key constraints, not sure for primary keys and unique constraints - see my comments above.

bricelam replied:
Fine as is--the earlier we throw the better.

ajcvickers commented:
@Brar Just wanted to let you know that we haven't forgotten about this. However, due to other high-priority work it make take a bit of time for us to get back to this.

Brar commented:
@ajcvickers thanks, no worries.
I'll try to come up with a suggestion addressing the comments.
We can pick up the discussion at that point as soon as you get to it.

bricelam commented:
Also remember that today they generate drop and re-create operations so they would be no more broken than they already are. We have a separate issue to perform table and constraint rebuilds automatically as part of a migration.

Brar commented:
Ok, this is my suggestion for further discussions.
There are still some areas in the code, where I'm uncertain, if my chosen solution is the best way to do it or even appropriate.
I've commented them in the code but I'll also list them here so that they don't get lost.

1. I had to move an `IMigrationsAnnotationProvider` down one level in the inheritance chain to [src/EFCore.Relational/Migrations/MigrationsSqlGenerator.cs](https://github.com/aspnet/EntityFrameworkCore/pull/14988/files#diff-feb942230d2c5a19a87bd4f62adb60a4) in order to be able to create annotations at this level. Please also look at the default instance and its `MigrationsAnnotationProviderDependencies`.
2. The default implementation of `RenameForeignKeyOperation` in [src/EFCore.Relational/Migrations/MigrationsSqlGenerator.cs](https://github.com/aspnet/EntityFrameworkCore/pull/14988/files#diff-feb942230d2c5a19a87bd4f62adb60a4) will reset the `OnUpdate` `ReferentialAction` to its default value.
3. I decided to still throw from the `Generate` overrides for the rename operations in [src/EFCore.Sqlite.Core/Migrations/SqliteMigrationsSqlGenerator.cs](https://github.com/aspnet/EntityFrameworkCore/pull/14988/files#diff-25d320eec09830c352a259b55e3dbf5e) because Sqlite doesn't currently support the add operations for primary/foreign keys and unique constraints so drop/create isn't an option here until https://github.com/aspnet/EntityFrameworkCore/issues/329 has landed.
4. I'm uncertain if the exception message thrown from the `Generate` methods for the rename operations in [src/EFCore.Relational/Migrations/MigrationsSqlGenerator.cs](https://github.com/aspnet/EntityFrameworkCore/pull/14988/files#diff-feb942230d2c5a19a87bd4f62adb60a4) is the right one.

Brar commented:
@ajcvickers @bricelam Now that 3.1 is out and you are beginning to work on 5.0 I assume that it's a good time for me to lobby for this PR again.
I'm willing to rebase it once again if there is time and dedication to review and possibly merge it in a timely manner.
Please go ahead and tell me if I should wait a litte and when it would best fit into your schedule. I just want to avoid working into the void.

cc: @roji

bricelam commented:
I hope to get this in before the first preview of of 5.0.0. It might be early January before I get around to it. (I'm currently writing [some docs](https://github.com/aspnet/EntityFramework.Docs/pull/1915) to support the 3.1 release.)

Brar commented:
Thanks @bricelam!
No hurry.
I'll rebase it in the next few weeks so that it's ready for re-review by January.

Brar commented:
@bricelam I've rebased this as I had some spare time today.
Don't feel pressured - January is fine!

bricelam commented on src/EFCore.SqlServer/Migrations/SqlServerMigrationsSqlGenerator.cs:
```diff
@@ -374,6 +372,29 @@ protected override void Generate(MigrationOperation operation, IModel model, Mig
+            Rename(
+                Dependencies.SqlGenerationHelper.DelimitIdentifier(operation.Name, operation.Schema),
+                operation.NewName,
+                "OBJECT",
```
OBJECT is the default. No need to specify it.

bricelam replied:
(same below)

Brar replied:
✔️

bricelam commented on src/EFCore.SqlServer/Migrations/SqlServerMigrationsSqlGenerator.cs:
```diff
@@ -48,7 +46,7 @@ public class SqlServerMigrationsSqlGenerator : MigrationsSqlGenerator
             [NotNull] MigrationsSqlGeneratorDependencies dependencies,
             [NotNull] IMigrationsAnnotationProvider migrationsAnnotations)
-            : base(dependencies) => _migrationsAnnotations = migrationsAnnotations;
+            : base(dependencies, migrationsAnnotations) { }
```
(Same for SQLite)

bricelam replied:
``` csharp
: base(dependenies.With(migrationsAnnotations))
```

Brar replied:
✔️

roji commented on test/EFCore.Relational.Specification.Tests/MigrationSqlGeneratorTestBase.cs:
```diff
@@ -387,6 +387,88 @@ public virtual void RenameTableOperation()
                 });
 
+        [ConditionalFact]
+        public virtual void RenameForeignKeyOperation()
```
I think these should be redone in the new functional MigrationsTestBase (which actually run against the database).

Brar replied:
This is a little more work.
I have to understand what needs to be done and implement it.
I'll have a look at this tomorrow.

Brar replied:
I think I've moved everything to it's proper place regarding the new test infrastructure.
I've also moved the tests in `EFCore.Relational.Tests`, `EFCore.Sqlite.FunctionalTests` and `EFCore.SqlServer.FunctionalTests` although, as I understand it, they aren't necessary any more.

@roji can you please have a look at the tests?

Brar commented on src/EFCore.Relational/Internal/DeleteBehaviorExtensions.cs:
```diff
@@ -0,0 +1,26 @@
+    {
+        public static ReferentialAction ToReferentialAction(this DeleteBehavior deleteBehavior)
+        {
+            switch (deleteBehavior)
```
✔️

Brar commented on src/EFCore.Relational/Migrations/Internal/MigrationsModelDiffer.cs:
```diff
@@ -1097,7 +1101,35 @@ protected virtual IEnumerable<MigrationOperation> Remove([NotNull] IProperty sou
+
+            if (sourceName != targetName)
+            {
+                if (target.IsPrimaryKey())
```
✔️

Brar commented on src/EFCore.Relational/Migrations/MigrationsSqlGenerator.cs:
```diff
@@ -78,18 +81,30 @@ private static readonly
+        /// <param name="migrationsAnnotations"> Provider-specific Migrations annotations to use. </param>
+        public MigrationsSqlGenerator(
+            [NotNull] MigrationsSqlGeneratorDependencies dependencies,
+            [CanBeNull] IMigrationsAnnotationProvider migrationsAnnotations = null)
```
Ok.
Thanks for the hint. It actually made me understand the API better.
✔️

Brar commented on src/EFCore.Relational/Migrations/MigrationsSqlGenerator.cs:
```diff
@@ -335,6 +350,67 @@ protected virtual IUpdateSqlGenerator SqlGenerator
+                    // Generate(RenameIndexOperation operation, IModel model, MigrationCommandListBuilder builder)
+                    // which served as template for this implementation but I actually also find the message
+                    // misleading/wrong in the original context as it is rather a failure while trying to look up
+                    // the foreign key (that IMO shouldn't happen) than an unsupported operation
```
✔️

Brar commented on src/EFCore.Relational/Migrations/MigrationsSqlGenerator.cs:
```diff
@@ -335,6 +350,67 @@ protected virtual IUpdateSqlGenerator SqlGenerator
+                PrincipalTable = foreignKey.PrincipalEntityType.GetTableName(),
+                PrincipalColumns = foreignKey.PrincipalKey.Properties.Select(p => p.GetColumnName()).ToArray(),
+                OnDelete = foreignKey.DeleteBehavior.ToReferentialAction(),
+                //OnUpdate = Whatever it was, we've reset it to default now
```
I think a warning is a good idea. I don't know the API for warnings though. I've found some warning related strings in the comments of `src/EFCore.Relational/Properties/RelationalStrings.resx`. Can you confirm that this is the right direction for my investigation of how to implement this?
🧐

> We should also add a TODO comment referencing issue #4073.

✔️

bricelam replied:
I'm not sure if we've generated warnings from here yet. I'll look into it.

Brar replied:
Ok, I've given my best to log a warning if we are going to loose the referential action.
I must admit though, that the logging system is pretty complicated and that I'm uncertain if I did everything the way it should be done.
Also I don't like my naming of the event, but there seems to be quite a bit of convention around logging, to which I tried to adhere.
I've left all related changes in a separate commit (453cfe00aed25e84645a03bd2ea64682b96b9357) to make reviewing them easier.

Brar commented:
> Some tests are failing (at the very least some SQL baselines need updating)

Sorry this was my fault.
After 3 Visual Studio crashes while trying to run more then one test at once, I decided to push this and hope for the best, which - as we all know - never goes well.

> I think we can drop the ModelDiffer and MigrationSqlGenerator tests

I haven't removed anything yet, but I can certainly do so. Just waiting for @bricelam to confirm that he's OK with that.

> If we're adding rename support for constraints, we may as well do the same for checked constraints

👍

bricelam commented:
Sounds good to me

Brar commented:
Hi @bricelam and @roji 
sorry for the delay.
A skiing-holiday and my day-job threw me back a little but now I think I've addressed all your feedback.

While most of the changes are pretty straightforward, I urge you to have a closer look at the last commit regarding the warning.
I'm by no means confident, that what I did there is the way it should be done.

smitpatel commented:
> skiing-holiday

@Brar - Skiiing always takes priority over anything else.
<img src="https://www.thisiscolossal.com/wp-content/uploads/2016/02/ski-spin.gif" />

Brar commented:
Ok @bricelam, I"ve rebased this once again.
Actually I've deemed it ready for another review since Feb 22 but I'm not sure that this has really come across.
As someone told me, the logging part is a no-go-area that only  @ajcvickers dares to enter so he might want to have a look at the last commit where I attempt to do something I don't completely understand.

Brar commented:
Since this has been inactive for quite a while and I don't think I can resume working on this any time soon I'm closing this now.
